File: Project2/neuralnetwork.cpp
#include <iostream>
#include <vector>
#include <random>
#include <cmath>

using namespace std;

typedef float (*func)(float);
typedef float (*func2)(float, float);

float myTanh(float x)
{
  return tanh(x);
}

float derivMyTanh(float x)
{
  float t = tanh(x);
  return 1 + t*t;
}

float sig(float x)
{
  return 1/(1 + exp(-x));
}

float derivSig(float x)
{
  float e = exp(-x);
  return e/((1 + e)*(1 + e));
}

float crossEntropy(float y_pred, float y)
{
  return (y_pred - y) / (y_pred * (1 - y_pred));
}


class NN
{

public:

  vector<int> dim;
  vector<func> acf;
  vector<func> dacf;

  int length;

  float*** weights;
  float** bias;

  float** z;
  float** a;

  float** grad;
  float*** delta;
  func2 cost;

  NN(vector<int> _dim, vector<func> _acf, vector<func> _dacf, func2 _cost)
  {
    dim = _dim;
    acf = _acf;
    dacf = _dacf;
    cost = _cost;
    length = dim.size()-1;
    default_random_engine generator;
    normal_distribution<float> distribution(0, 1);

    weights = new float**[length];
    bias = new float*[length];
    z = new float*[length];
    a = new float*[length];
    grad = new float*[length];
    delta = new float**[length];

    for(int i = 0; i<length; i++)
    {
      int m = dim[i+1];
      int n = dim[i];
      weights[i] = new float*[m];
      delta[i] = new float*[m];
      bias[i] = new float[m];
      z[i] = new float[m];
      a[i] = new float[m];
      grad[i] = new float[m];

      for(int j = 0; j<m; j++)
      {
        bias[i][j] = 0.01;
        z[i][j] =  a[i][j] = 0;

        weights[i][j] = new float[n];
        delta[i][j] = new float[n];
        for(int k = 0; k<n; k++)
        {
          weights[i][j][k] = distribution(generator);
        }
      }
    }
  }

  void forward(float* x)
  {
    int m = dim[1];
    int n = dim[0];
    for(int i=0; i<m; i++)
    {
      z[0][i] = bias[0][i];
      for(int j=0; j<n; j++)
      {
        z[0][i] += weights[0][i][j]*x[j];
      }
    }

    for(int i=1; i<length; i++)
    {
      int m = dim[i+1];
      int n = dim[i];

      for(int j=0; j<m; j++)
      {
        z[i][j] = bias[i][j];
        a[i][j] = 0;
        for(int k=0; k<n; k++)
        {
          z[i][j] += weights[i][j][k]*a[i-1][k];
        }
        a[i][j] = acf[i-1](z[i][j]);
      }
    }
  }

  void backward(float* x, float* y)
  {
    forward(x);
    int m;
    int n = dim[length];

    for(int i=0; i<n; i++)
    {
      grad[length-1][i] = dacf[length-1](z[length-1][i])*cost(a[length-1][i], *y);
    }

    for(int i=length-1; i>0; i--)
    {
      m = dim[i+1];
      n = dim[i];

      for(int j=0; j<n; j++)
      {
        grad[i-1][j] = 0;
        for(int k=0; k<m; k++)
        {
          grad[i-1][j] += weights[i][k][j]*grad[i][k];
        }
        grad[i-1][j] *= dacf[i-1](z[length-1][j]);
      }
    }
  }

    void train(float** x, float** y, int n)
    {
        for(int i=0; i<n; i++)
        {
            for(int j=0; j<dim.size()-1; j++)
            {
                int m = dim[j+1];
                int n = dim[j];
                for(int k=0; k<m; k++)
                {
                    for(int l=0; l<n; l++)
                    {
                        delta[j][k][l] = 0;
                    }
                }
            }
        }

        for(int i=0; i<n; i++)
        {
            backward(x[0], y[0]);
            for(int j=0; j<dim.size()-1; j++)
            {
                int m = dim[j+1];
                int n = dim[j];
                for(int k=0; k<m; k++)
                {
                    for(int l=0; l<n; l++)
                    {
                        delta[j][k][l] += grad[j][k]*a[j][l];
                    }
                }
            }
        }

        for(int j=0; j<dim.size()-1; j++)
        {
            int m = dim[j+1];
            int n = dim[j];
            for(int k=0; k<m; k++)
            {
                for(int l=0; l<n; l++)
                {
                    weights[j][k][l] -= 0.001*delta[j][k][l];
                }
                bias[j][k] -= 0.001*grad[j][k];
            }
        }
    }
};

int main()
{
    vector<int> dim{2, 4, 1};
    vector<func> acf{myTanh, sig};
    vector<func> dacf{derivMyTanh, derivSig};

    NN nn(dim, acf, dacf, crossEntropy);


    float** x = new float*[1];
    float** y = new float*[1];
    x[0] = new float[2];
    y[0] = new float[1];
    x[0][0] = 1; x[0][1] = 2;
    y[0][0] = 2;

    int n = 2000;
    nn.forward(x[0]);
    cout << nn.a[1][0] << endl;
    for(int i=0; i<n; i++)
    {
        nn.train(x, y, 1);
    }

    nn.forward(x[0]);
    cout << nn.a[1][0] << endl;


return 0;
}

File: Project2/src/neuralnetwork.py
from urllib.request import urlopen
import numpy as np
from sklearn.datasets import load_digits
import sys

import pickle
import os
os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'


class Sigmoid():
    def __call__(self, x):
        return 1 / (1 + np.exp(-x))

    def deriv(self, x):
        return np.exp(-x) / (1 + np.exp(-x))**2


class Tanh():
    def __call__(self, x):
        return np.tanh(x)

    def deriv(self, x):
        return 1 + np.tanh(x)**2


class Relu():
    def __call__(self, x):
        return np.maximum(x, 0)

    def deriv(self, x):
        return (0 < x).astype("int")


class Pass():
    def __call__(self, x):
        return x

    def deriv(self, x):
        return 1


class SquareLoss():
    def __call__(self, y_pred, y):
        return 0.5 * sum((y_pred - y)**2)

    def deriv(self, y_pred, y):
        return y_pred - y


class CrossEntropy():
    def __call__(self, y_pred, y):
        return -sum(y * np.log(y_pred) + (1 - y) * log(1 - y_pred))

    def deriv(self, y_pred, y):
        return (y_pred - y) / (y_pred * (1 - y_pred))


class NeuralNetwork():

    def __init__(self, dim, acf, cost):
        self.dim = dim
        self.acf = np.array(acf)
        self.cost = cost

        self.W = np.empty(len(dim) - 1, dtype=np.ndarray)  # weight matricies
        self.b = np.empty(len(dim) - 1, dtype=np.ndarray)  # biases
        self.z = np.empty(len(dim), dtype=np.ndarray)
        self.a = np.empty(len(dim), dtype=np.ndarray)
        self.grad = np.empty(len(dim) - 1, dtype=np.ndarray)
        self.delta = np.empty(len(dim) - 1, dtype=np.ndarray)

        for i in range(len(dim) - 1):
            m = dim[i + 1]
            n = dim[i]
            self.W[i] = np.random.normal(0, 1, (m, n))
            self.b[i] = 0.01 * np.ones(m)

    def forward(self, x):
        self.z[0] = x
        self.a[0] = x
        for i in range(len(self.W)):
            self.z[i + 1] = self.a[i]@self.W[i].T + self.b[i][np.newaxis]
            self.a[i + 1] = self.acf[i](self.z[i + 1])

    def backward(self, x, y):
        self.forward(x)

        self.delta[-1] = self.acf[-1].deriv(self.z[-1]) * \
            self.cost.deriv(self.a[-1], y)

        for i in range(len(self.W) - 1, 0, -1):
            self.delta[i - 1] = self.delta[i] @ self.W[i] * \
                self.acf[i - 1].deriv(self.z[i])

    def train(self, X, y, mu, lamb, batch_size, epochs):
        if len(y.shape) == 1:
            y = y[:, np.newaxis]

        n = len(y)
        num_iters = int(n / batch_size)

        for e in range(epochs):

            if e % (epochs / 100) == 0:
                sys.stdout.write("\r" + "%d" % (100 * e / epochs))
                sys.stdout.flush()

            for i in range(num_iters):
                idx_train = np.random.choice(
                    np.arange(0, n), batch_size, replace=False)
                self.backward(X[idx_train], y[idx_train])
                for j in range(len(self.grad)):
                    self.grad[j] = self.delta[j].T @ self.a[j]

                self.W -= mu * self.grad + lamb * self.W

                for j in range(len(self.grad)):
                    self.b[j] -= mu * np.sum(self.delta[j], axis=0)

    def predict(self, X):
        self.forward(X)
        return self.a[-1]


if __name__ == "__main__":
    tanh = Tanh()
    sig = Sigmoid()
    softMax = SoftMax()
    relu = Relu()

    crossEntropy = CrossEntropy()
    squareLoss = SquareLoss()
    mlogloss = MlogLoss()

    url_main = "https://physics.bu.edu/~pankajm/ML-Review-Datasets/isingMC/"
    data_file_name = "Ising2DFM_reSample_L40_T=All.pkl"
    label_file_name = "Ising2DFM_reSample_L40_T=All_labels.pkl"

    print("1")

    data = pickle.load(urlopen(url_main + data_file_name))
    print("2")
    data = np.unpackbits(data).reshape(-1, 1600)
    print("3")
    data = data.astype('int')

    print(data.shape)
